### ISSUE TITLE
rustbuild do not delete destination directory

### ISSUE THREAD

**Issue by nagisa** [rust-lang/rust#34279]
I tend to put various test files into `$DESTINATION/stage1/bin` so I can have a different terminal open and quickly test stage1 rustc against various test cases I have at hand for a particular thing I’m working on at the time without cluttering root of my checkout.

rustbuild deleting the destination directory every build is inherently incompatible with this workflow. Consider not doing it.

**Comment by alexcrichton** [comment]
Yeah right now this is done to ensure we clean everything out, but this definitely bit Cargo in the past so we should probably figure our a more principled way of doing this.

**Comment by jyn514** [comment]
Triage: this is still an issue: https://github.com/rust-lang/rust/blob/d50583eaecdf88b5ee59e1d77fe3b4a8d87cbe6e/src/bootstrap/compile.rs#L1046

**Comment by jyn514** [comment]
I think this is hard to do in a way that's not surprising. Consider the following invocations:
1. `rustdoc +stage1 --version`: `error: 'rustdoc' is not installed for the toolchain 'stage1'` (good, it's never been built so we should error)
2. `x build --stage 1 && rustdoc +stage1 --version`: `rustdoc 1.63.0-dev` (good, we just built it)
3. some edit to compiler/
4. `x build --stage 1 library/std && rustdoc +stage1 --version`: `rustdoc 1.63.0-dev` (confusing, this is using an outdated version of rustdoc that doesn't have your latest edits

I suppose we could only delete files we know that bootstrap generates, but that seems quite difficult to maintain ...

**Comment by jyn514** [comment]
RalfJung [mentions](https://github.com/rust-lang/rust/issues/53650#issuecomment-1364698742) that this makes it annoying (and due to #90224, currently impossible) to have both `rustc_private` and `rustdoc` artifacts in the same sysroot.

davidtwco [mentions](https://github.com/rust-lang/rust/issues/52565) that this means that running `x check` between invocations of `x build` breaks `--keep-stage`.
